### PR TITLE
R python plugin

### PR DESCRIPTION
**Pull request by myyong**
@tpetricek I've made the rPythonPlugin a class rather than a const (https://github.com/wrattler/wrattler/blob/rPythonPlugin/client/src/languages/rPython.ts#L71). Are you happy with this?

### CODE REVIEW

File: client/src/definitions/languages.ts [hunk 1]
```diff
@@ -29,6 +29,9 @@ interface LanguagePlugin {
   /** Returns a language-specific editor that handles the UI in a notebook  */
   editor : Editor<EditorState, any>
 
+  /** Returns a language-specific editor that handles the UI in a notebook  */
+  serviceURI? : string
```
Comment (tpetricek) [inline client/src/definitions/languages.ts]: I don't think this should be here - the fact that the language uses a service should be completely hidden in the implementation. The system should not need to know about this fact.

File: client/src/languages/rPython.ts [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+declare var DATASTORE_URI: string;
+
+// ------------------------------------------------------------------------------------------------
+// rPython plugin
+// ------------------------------------------------------------------------------------------------
+
+/// A class that represents a Python block. All blocks need to have 
+/// `language` and Python also keeps the Python source we edit and render
+
+type languageType = "python" | "r"
```
Comment (tpetricek) [inline client/src/languages/rPython.ts]: We should not restrict this to R and Python - anybody who can provide a URL for another service should be able to use this class (without modifying it) to create a new language plugin for their language.

File: client/src/languages/rPython.ts [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+
+export class rPythonBlockKind implements Langs.Block {
```
Comment (tpetricek) [inline client/src/languages/rPython.ts]: Again, I think this should not be just `rPythonBlockKind`. We need a better name.

 - Maybe `LanguageServiceBlockKind` (but that's a bit long)
 - or `ApiBlockKind` (but that's not very clear)
 - or maybe `ExternalBlockKind` ?

If you have better ideas, please share them! I think I like `ExternalBlockKind` the most out of the above.

File: client/src/languages/rPython.ts [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+				importedFrames.push({ name: imported.variableName, url: (<Values.DataFrame>imported.value).url })
+				}
+						let src = pyNode.source
+						let hash = Md5.hashStr(src)
+						let body = {"code": src,
+										"hash": hash,
+										"frames": importedFrames}
+						return await getEval(body, this.serviceURI);
+			case 'export':
+				let pyExportNode = <Graph.PyExportNode>node
```
Comment (tpetricek) [inline client/src/languages/rPython.ts]: We should also rename `PyExportNode` to `ExternalExportNode`.

Comment (tpetricek) [inline client/src/languages/rPython.ts]: We should also rename `PyExportNode` to `ExternalExportNode`.

File: client/src/languages/rPython.ts [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+	parse (code:string) {	
+		return new rPythonBlockKind(code, this.language);
+	}
+
+	bind (scopeDictionary: Langs.ScopeDictionary, block: Langs.Block) {
+			let pyBlock = <rPythonBlockKind>block
+			// return getCodeExports(scopeDictionary, pyBlock.source)
+			let dependencies:Graph.PyExportNode[] = [];
+			let node:Graph.PyCodeNode = {
+				language:"python",
```
Comment (tpetricek) [inline client/src/languages/rPython.ts]: This should be the language provided when constructing the plugin - not `python`.

File: client/src/languages/rPython.ts [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+// 					}
+// 				}
+// 				return {code: node, exports: dependencies};
+// 			} catch (error) {
+// 				console.error(error);
+// 			}
+// 		}
+// 		return getExports()
+// 	}
+// }
```
Comment (tpetricek) [inline client/src/languages/rPython.ts]: Can you delete all the commented-out code? If we ever need it, we have it somewhere in the Git history!

File: server/R/.Rhistory [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+r$content
+r$status_code
+content(r)
+content(r,"json")
+content(r,"parsed")
+library(xml2)
+install.packages("xml2")
+install.packages("xml2")
+devtools::install_github("r-lib/xml2")
+quit()
```
Comment (tpetricek) [inline server/R/.Rhistory]: Can we remove this file from the PR?

**Comment by tpetricek** [comment]
I added a couple of comments for improving the PR. 

Aside from the above, I would also like to consistently use spaces everywhere in the code - I think having tabs is making merging and reviewing things on GitHub really hard. I don't particularly mind if we use 2 or four spaces, but it would be good to be consistent.

File: client/src/languages/rPython.ts (hunk 2 above)
Comment (myyong) [reply client/src/languages/rPython.ts]: agree, I will remove this

File: client/src/languages/rPython.ts (hunk 3 above)
Comment (myyong) [reply client/src/languages/rPython.ts]: Yes, Nick and I were talking about it. We weren't happy with rPythonPlugin either. We thought apiBasedLanguagePlugin, but then Gamma is api based isn't it. I'll keep this open for more suggestions.

File: client/src/languages/external.ts [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+import axios from 'axios';
+
+declare var DATASTORE_URI: string;
+
+// ------------------------------------------------------------------------------------------------
+// rPython plugin
+// ------------------------------------------------------------------------------------------------
+
+/// A class that represents a Python block. All blocks need to have 
+/// `language` and Python also keeps the Python source we edit and render
```
Comment (tpetricek) [inline client/src/languages/external.ts]: The above comment also needs updating - this is not R/Python anymore.

File: client/src/languages/javascript.ts [hunk 9]
```diff
@@ -85,39 +99,49 @@ class JavascriptBlockKind implements Langs.Block {
...
   const javascriptEditor : Langs.Editor<JavascriptState, JavascriptEvent> = {
     initialize: (id:number, block:Langs.Block) => {  
-      return { id: id, block: <JavascriptBlockKind>block}
+      return { id: id, block: <JavascriptBlockKind>block, tabID:0}
     },
   
-    update: (state:JavascriptState, event:JavascriptEvent) => state,
+    update: (state:JavascriptState, event:JavascriptEvent) => {
+      switch(event.kind) {
+				case 'switchtab':
```
Comment (tpetricek) [inline client/src/languages/javascript.ts:110]: There are a few more tabs left in this file.

File: client/src/languages/javascript.ts [hunk 10]
```diff
@@ -85,39 +99,49 @@ class JavascriptBlockKind implements Langs.Block {
...
+    //       }, 
+    //       [ (cell.code.value==undefined) ? evalButton : ("Value is: " + JSON.stringify(cell.code.value)) ]),
+    //   ]);
  
-			let code = createEditor("javascript", state.block.source, cell, context)
-      return h('div', { }, [code, results])
-    },
+		// 	let code = createEditor("javascript", state.block.source, cell, context)
+    //   return h('div', { }, [code, results])
+    // },
```
Comment (tpetricek) [inline client/src/languages/javascript.ts]: If this is not needed, then please delete commented out code.

File: client/src/languages/external.ts (hunk 8 above)
Comment (tpetricek) [inline client/src/languages/external.ts]: Can we please update the above comment to be up-to-date? It's not just R/Python anymore...

**Comment by tpetricek** [comment]
This is looking great now! 

@myyong I added three more comments, but they are all very minor details. If you could look at them, that would be great, and let's merge this.

Nice work!

File: client/src/languages/javascript.ts (hunk 9 above)
Comment (myyong) [reply client/src/languages/javascript.ts:110]: Removed in commit https://github.com/wrattler/wrattler/commit/baf798530a21ccb8198bd9930f199f68f034b553

**Comment by tpetricek** [comment]
👍

**Pull request merged**